httpx/middleware/auth.py

```diff
@@ -0,0 +1,200 @@
+import hashlib
+import os
+import time
+import typing
+from base64 import b64encode
+
+from ..exceptions import ProtocolError
+from ..models import AsyncRequest, AsyncResponse, StatusCode
+from ..utils import to_bytes, to_str, unquote
+from .base import BaseMiddleware
+
+
+class BasicAuthMiddleware(BaseMiddleware):
+    def __init__(
+        self, username: typing.Union[str, bytes], password: typing.Union[str, bytes]
+    ):
+        if isinstance(username, str):
+            username = username.encode("latin1")
+
+        if isinstance(password, str):
+            password = password.encode("latin1")
+
+        userpass = b":".join((username, password))
+        token = b64encode(userpass).decode().strip()
+
+        self.authorization_header = f"Basic {token}"
+
+    async def __call__(
+        self, request: AsyncRequest, get_response: typing.Callable
+    ) -> AsyncResponse:
+        request.headers["Authorization"] = self.authorization_header
+        return await get_response(request)
+
+
+class CustomAuthMiddleware(BaseMiddleware):
+    def __init__(self, auth: typing.Callable[[AsyncRequest], AsyncRequest]):
+        self.auth = auth
+
+    async def __call__(
+        self, request: AsyncRequest, get_response: typing.Callable
+    ) -> AsyncResponse:
+        request = self.auth(request)
+        return await get_response(request)
+
+
+class HTTPDigestAuthMiddleware(BaseMiddleware):
+
+    ALGORITHM_TO_HASH_FUNCTION: typing.Dict[str, typing.Callable] = {
+        "MD5": hashlib.md5,
+        "MD5-SESS": hashlib.md5,
+        "SHA": hashlib.sha1,
+        "SHA-SESS": hashlib.sha1,
+        "SHA-256": hashlib.sha256,
+        "SHA-256-SESS": hashlib.sha256,
+        "SHA-512": hashlib.sha512,
+        "SHA-512-SESS": hashlib.sha512,
+    }
+
+    def __init__(
+        self, username: typing.Union[str, bytes], password: typing.Union[str, bytes]
+    ) -> None:
+        self.username = to_bytes(username)
+        self.password = to_bytes(password)
+        self._previous_nonce: typing.Optional[bytes] = None
+        self._nonce_count = 0
+
+    async def __call__(
+        self, request: AsyncRequest, get_response: typing.Callable
+    ) -> AsyncResponse:
+        response = await get_response(request)
+
+        if self._should_return_digest_auth(response):
+            request.headers["Authorization"] = self._build_auth_header(
+                request, response
+            )
+            return await self(request, get_response)
+
+        return response
+
+    def _should_return_digest_auth(self, response: AsyncResponse) -> bool:
+        auth_header = response.headers.get("www-authenticate")
+        return StatusCode.is_client_error(response.status_code) and (
+            auth_header is None or "digest" in auth_header.lower()
+        )
```

Comment:
What happens if we send the authorization header after receiving the first 401, but it is wrong (eg wrong username and password)?

The ‘is None’ check part makes me think that we’d go in an infinite loop: we would build and send the authorization header again, and fail, build and send it, and so on...

I suppose that the 401 if we pass wrong credentials won’t contain the WWW-Authenticate header, so should we instead check that it is present and digest is in there?

Reply:
That's a really good point, as it stands we enter an infinite loop since the server keeps returning 401 with a different nonce which we are not handling. 👍